add endstop monitor to M43 command

Bob-the-Kuhn opened the pull request:
Added a debug/setup aid that I call endstop monitor.

This routine monitors the endstops and the probe and, if a change is
detected, sends a message to the host and toggles the LED.

This routine runs in the background so normal machine operation is not
affected.

M43 E toggles the function on & off.

The main routine was added to the temperature ISR.  The call to the
routine mimics a slow PWM.  It gets called at a 15Hz rate.

It adds 1 global variable and two static variables.

I also updated the pin list in pinsDebug.h.  All the previous pins are
still present.  I took a copy of RCBugFix from 7 Oct, did a grep to get
all the #defines, kept the ones that had a pin, added in all the
previous pins, & then used Excel to create the "#if pin ... print pin
... #endif" sequence.

I'm not sure what the intent of the pin list in pinsDebug.h so just
ignore my changes if they conflict with something.

I also added M43 to the list of implemented commands in Marlin_main.cpp

Bob-the-Kuhn committed:
added endstop monitor to M43 command

Added a debug/setup aid that I call endstop monitor.

This routine monitors the endstops and the probe and, if a change is
detected, sends a message to the host and toggles the LED.

This routine runs in the background so normal machine operation is not
affected.

M43 E toggles the function on & off.

The main routine was added to the temperature ISR.  IThe call to the
routine mimics a slow PWM.  It gets called at a 15Hz rate.

It adds 1 global variable and two static variables.

I also updated the pin list in pinsDebug.h.  All the previous pins are
still present.  I took a copy of RCBugFix from 7 Oct, did a grep to get
all the #defines, kept the ones that had a pin, added in all the
previous pins, & then used Excel to create the "#if pin ... print pin
... #endif" sequence.

I'm not sure what the intent of the pin list in pinsDebug.h so just
ignore my changes if they conflict with something.

I also added M43 to the list of implemented commands in Marlin_main.cpp

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4734,9 +4746,15 @@ inline void gcode_M42() {
+    
+    if (code_seen('E')) {
+      endstop_monitor_flag = endstop_monitor_flag ? false : true;
+      SERIAL_PROTOCOLLN((endstop_monitor_flag ? "endstop monitor active" : "endstop monitor disabled"));
```
Size is more important than performance here.
To optimize string storage / program size, change to:

``` cpp
SERIAL_PROTOCOLPGM("endstop monitor ");
SERIAL_PROTOCOL(endstop_monitor_flag ? "en" : "dis");
SERIAL_PROTOCOLLNPGM("abled");
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4734,9 +4746,15 @@ inline void gcode_M42() {
+          report_pin_state(pin);
+    
+    if (code_seen('E')) {
+      endstop_monitor_flag = endstop_monitor_flag ? false : true;
```
``` cpp
endstop_monitor_flag ^= true;
```

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1390,8 +1395,91 @@ void Temperature::set_current_temp_raw() {
+      if (current_endstop_bits_local != old_endstop_bits_local) {
+        analogWrite(LED_PIN, local_LED_status );                // toggle LED
+        SERIAL_PROTOCOLPGM("\n");                               // make it easy to see the message
+        SERIAL_PROTOCOLPGM("\n");
```
`SERIAL_PROTOCOLPGM("\n\n")` is fine here.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1390,8 +1395,91 @@ void Temperature::set_current_temp_raw() {
+   * Yes, we could miss a rapid back & forth change but
+   * that won't matter because this is all manual.
+   *
+   */ 
```
Comments like this should be placed above the function definition rather than within it.

thinkyhead commented on Marlin/pinsDebug.h:
```diff
@@ -57,372 +57,598 @@ static bool report_pin_name(int8_t pin) {
+  #if defined(X_STEP_PIN) && X_STEP_PIN > -1
+    PIN_SAY(X_STEP_PIN)
   #endif
+  #if defined(X_STOP_PIN) && X_STOP_PIN > -1
```
When a pin name actually ends with `_PIN` you should use the `PIN_EXISTS` macro. In this case `PIN_EXISTS(X_STOP)`. For the "`STOP`" pins, also check whether they are equal to the `MIN` or `MAX` pins, and don't bother displaying them if they're assigned so.

Bob-the-Kuhn replied:
I disagree about the "STOP" pins.  Reporting that they exist shouldn't be a problem.  Keeping people from writing to them is a must.

The "don't report protected pins" function is already implemented in the M43 command.  It checks if the pin is protected before calling the "report_pin_name" function so the M43 command won't print out anything about the endstops.

thinkyhead commented on Marlin/pinsDebug.h:
```diff
@@ -57,372 +57,598 @@ static bool report_pin_name(int8_t pin) {
-  #if PIN_EXISTS(SERVO0)
-    PIN_SAY(SERVO0_PIN);
+  #if defined(__FD) && __FD > -1   // Pin list updated from 7 OCT RCBugfix branch
+    PIN_SAY(__FD)
```
All the `PIN_SAY` lines should end with a semicolon. To get rid of compiler complaints, remove the semicolon I accidentally left at the end of the `PIN_SAY` macro definition.

Bob-the-Kuhn replied:
Did the same thing for ANALOG_PIN_SAY

Bob-the-Kuhn commented on Marlin/pinsDebug.h:
```diff
@@ -57,372 +57,598 @@ static bool report_pin_name(int8_t pin) {
+  #if defined(X_STEP_PIN) && X_STEP_PIN > -1
+    PIN_SAY(X_STEP_PIN)
   #endif
+  #if defined(X_STOP_PIN) && X_STOP_PIN > -1
```
I disagree about the "STOP" pins.  Reporting that they exist shouldn't be a problem.  Keeping people from writing to them is a must.

The "don't report protected pins" function is already implementd in the M43 command.  It checks if the pin is protected before calling the "report_pin_name" function so the M43 command won't print out anything about the endstops.

Bob-the-Kuhn committed:
cleanup of endstop monitor

Implemented everything requested by Scott except for the "don't report
protected pins" function.

In addition a)  I removed the ";" from the ANALOG_PIN_SAY macro
definition so the  ANALOG_PIN_SAY lines would end in a semicolon and b)
moved the static defines into the routine so that all the endstop
monitor items are in one place.

I don't see any problems with reporting that a protected pin exists.
Keeping someone from writing to them is a must.

The "don't report protected pins" function is already implemented in the
M43 command.  It checks if the pin is protected before calling the
"report_pin_name" function so the M43 command won't print out anything
about the endstops.

Bob-the-Kuhn commented:
Implemented everything requested by Scott except for the "don't report protected pins" function.

In addition a)  I removed the ";" from the ANALOG_PIN_SAY macro definition so the  ANALOG_PIN_SAY lines would end in a semicolon and b) in temperature.cpp moved the static defines into the routine so that all the endstop monitor items are in one place.

I don't see any problems with reporting that a protected pin exists.  Keeping someone from writing to them is a must.

The "don't report protected pins" function is already implemented in the M43 command.  It checks if the pin is protected before calling the "report_pin_name" function so the M43 command won't print out anything about the endstops.

thinkyhead commented:
> except for the "don't report protected pins" function

I didn't request this. I did request that you display the endstop pins only once instead of duplicating them with two different names, however. Perhaps that what came across incorrectly.

Bob-the-Kuhn commented:
Now I understand ...

There are 24 boards that use STOP instead of MAX/MIN. Pins.h assigns them to either MAX or MIN depending on the home direction.

I think it would be confusing for the owners of these boards to see a different name listed than what's in the pins_xxx file.

I can easily delete the STOP pins from the list if that's what you still want.

Bob-the-Kuhn commented:
We can stop scanning pins 70 and above.  The tables in pins_arduino.h for the mega only go from 0-69 so digitalRead just reports "0" for all pins above 69.

Bob-the-Kuhn committed:
removed STOP pins from list

commented out the STOP pins

Doesn't really matter because the corresponding MAX/MIN pin is
encountered first at which point the search stops.

Bob-the-Kuhn commented:
I've commented out the STOP pins and reduced the pins to 0-69.

Doesn't really matter because the corresponding MAX/MIN pin is encountered first at which point the search stops.

Bob-the-Kuhn committed:
M43  - made report pretty & added PWM info

pinsDebug changes
made things line up in columns
added extended report
PWM info added in the extended report

Marlin_main.cpp changes
changed one line in M43 routine
if no options selected now does extended report
added "hidden" option to make the extended report
ignore the protected status
brought up to date with 26 OCT RCBugFix

Bob-the-Kuhn commented:
**M43 - made report pretty & added PWM info** 

**pinsDebug** changes
    made things line up in columns
    added extended report
        PWM info added in the extended report

**Marlin_main.cpp** changes
    changed one line in M43 routine
        if no options selected now does extended report
            added "hidden" option to make the extended report ignore the protected status
   brought Marlin_main.cpp up to date with 26 OCT RCBugFix

Bob-the-Kuhn committed:
missed some updates to Marlin_main.cpp

Today's earlier update was done against an out of date copy of RCBugFix.

This one was done against 28 OCT 5:15 PM Central time zone.

Bob-the-Kuhn commented:
missed some updates to Marlin_main.cpp 

Today's earlier update was done against an out of date copy of RCBugFix.

This one was done against 28 OCT 5:15 PM Central time zone.

Bob-the-Kuhn commented:
Found a bug in the pinsDebug.h file.

I'll sit on it until this PR bubbles up on Scott's to-do-list for him to look at it.

Bob-the-Kuhn committed:
corrected register names & removed custom print code

1) corrected register names - had some "B" & "C" where "A" should have
been

2) replaced custom print code with sprintf statements

3) widened name column

Bob-the-Kuhn commented:
This PR has been replaced (cloned) by PR #5112
